File: 2019/B/stones.cc
#include <bits/stdc++.h>

using namespace std;

int T, N;
int S[100], E[100], L[100], P[100];
int cache[10000][100];

bool cmp(int i, int j) {
  return L[i] * S[j] > L[j] * S[i];
}

void init_p(int n) {
  for(int i=0; i < n; i++) {
    P[i] = i;
  }
}

int max_energy(int time, int i) {
  if(i >= N) return 0;
  if(cache[time][i] != -1) return cache[time][i];
  int j = P[i];
  int en_diff = E[j] - L[j] * time;
  int en = max(0, en_diff);
  int en1 = max_energy(time + S[j], i + 1) + en;
  int en2 = max_energy(time, i + 1);
  return cache[time][i] = max(en1, en2);
}

int main() {
  scanf("%d\n", &T);
  for(int t = 1; t <= T; t++) {
    scanf("%d", &N);
    for(int n = 0; n < N; n++) {
      scanf("%d%d%d", &S[n], &E[n], &L[n]);
    }
    init_p(N);
    memset(cache, -1, sizeof cache);
    sort(P, P + N, cmp);
    int e = max_energy(0, 0);
    printf("Case #%d: %d\n", t, e);
  }
}

File: 2019/B/stones.py
import numpy as np


def sort():
    rating = -L / S
    return np.argsort(rating)

def max_energy(time, i):
    if i >= N: return 0
    if Cache[time][i] != -1: return Cache[time, i];
    j = P[i]
    en_rem = E[j] - L[j] * time # remaining power at t=time
    en = max(0, en_rem) # powerup
    en1 = max_energy(time + S[j], i + 1) + en
    en2 = max_energy(time, i + 1)
    c =  max(en1, en2)
    Cache[time, i] = c
    return c

T = int(input())

for t in range(1, T + 1):
    N = int(input())
    S = np.empty(N, np.int_)
    E = np.empty(N, np.int_)
    L = np.empty(N, np.int_)
    Cache = np.full((100 * N, N), -1)
    for n in range(N):
        S[n], E[n], L[n] = map(int, input().split())
    P = sort()
    e = max_energy(0, 0)
    print("Case #{}: {}".format(t, e))
